src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,701 @@
+    pub const BYTEORDER: &str = "byteorder";
+}
+
+/// Name of the new empty cargo project to be created to run `cargo test -Z
+/// build-std` in.
+const EMPTY_PROJECT_NAME: &str = "mir-json-translate-libs-empty-project";
+
+/// Deserialized form of cargo --unit-graph JSON output.
+/// See https://doc.rust-lang.org/cargo/reference/unstable.html#unit-graph
+#[derive(Deserialize)]
+struct UnitGraph {
+    version: u32,
+    units: Vec<UnitGraphUnit>,
+    roots: Vec<usize>,
+}
+
+#[derive(Deserialize)]
+struct UnitGraphUnit {
+    pkg_id: PackageId,
+    target: Target,
+    dependencies: Vec<UnitGraphDependency>,
+}
+
+#[derive(Clone, Deserialize)]
+struct UnitGraphDependency {
+    index: usize,
+    extern_crate_name: CrateName,
+}
+
+/// Module to hide field of newtype [CrateName].
+mod crate_name {
+    use std::fmt;
+
+    use serde::Deserialize;
+
+    /// Newtype wrapper around [String] which enforces that all `-` are replaced
+    /// with `_`. Package names often contain `-`, but it is not a valid
+    /// character in Rust identifiers, and crate names must be valid Rust
+    /// identifiers since they are automatically exposed as module names in Rust
+    /// code. Therefore cargo converts `-` to `_` when deriving crate names from
+    /// package names.
+    ///
+    /// Construct with [From<&str>].
+    #[derive(Clone, Deserialize, PartialEq)]
+    pub struct CrateName(String);
+
+    impl From<&str> for CrateName {
+        fn from(name: &str) -> Self {
+            Self(name.replace("-", "_"))
+        }
+    }
+
+    impl From<CrateName> for String {
+        fn from(crate_name: CrateName) -> String {
+            crate_name.0
+        }
+    }
+
+    impl fmt::Display for CrateName {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            self.0.fmt(f)
+        }
+    }
+}
+use crate_name::CrateName;
+
+/// Our version of the unit graph, which allows for adding our custom libraries
+/// and easily converting to mir-json commands.
+struct CustomUnitGraph {
+    units: Vec<CustomUnit>,
+    roots: Vec<usize>,
+}
+
+struct CustomUnit {
+    target: CustomTarget,
+    dependencies: Vec<UnitGraphDependency>,
+}
+
+enum CustomTarget {
+    /// A library to be compiled for the target architecture. We really only
+    /// care about the target information for these, so only this variant has
+    /// fields.
+    TargetLib(CustomTargetLib),
+    /// A dependency to be compiled for the host, like a build script or proc
+    /// macro.
+    HostDep,
+    /// A binary that is not a [CustomTarget::HostDep].
+    FinalBin,
+}
+
+/// The fields here are a combination of the fields we care about from [Target],
+/// [cargo_metadata::Artifact], and [cargo_metadata::BuildScript]. Each
+/// corresponds to some input to `rustc`.
+#[derive(Clone)]
+struct CustomTargetLib {
+    /// `--crate-name`
+    crate_name: CrateName,
+    /// source file argument
+    src_path: Utf8PathBuf,
+    /// `--edition`
+    edition: Edition,
```

Comment:
Seeing this makes me wonder whether `--unit-graph` is the right thing to use for this tool.  The `--build-plan` output gives the full rustc arguments and environment for each step, whereas here it seems like we're reconstructing a bunch of that ourselves.

Reply:
It seems like `--build-plan` is [considered broken and will possibly be removed in the future](https://github.com/rust-lang/cargo/issues/7614). Also, when used in conjunction with `build-std`, Cargo outputs a warning that "`-Zbuild-std` does not fully support `--build-plan`". So I didn't use it.